### ISSUE TITLE
[BUG] UnboundLocalError in Python but not in Cython

### ISSUE THREAD

**Issue by Zabolekar** [cython/cython#3895]
Windows 10, Python 3.8.3, Cython 3.0a6, Visual Studio 2017 version 15.9.20.

Code example:

```python
def f(n):
    if n:
        b = 0
    print(b)

def g():
    f(0)
```

When imported as a pure Python module, `g()` raises an exception:

`UnboundLocalError: local variable 'b' referenced before assignment`

**Expected behavior**

When this code is compiled as a Cython extension and imported, `g()` behaves like in Python.

**Observed behavior**

When this code is compiled as a Cython extension and imported, `g()` prints some random int, like -874173976 or 24.

**Likely reason**

In the generated C file, `long __pyx_v_b` is declared right at the top of the function that corresponds to `f`. Actually, this looks intentional and I'm not sure whether it's really a bug. But I couldn't find anywhere it in the documentation.

**Comment by da-woods** [comment]
If you need to avoid this you can set `#cython: infer_types=False` as a [compiler directive](http://docs.cython.org/en/latest/src/userguide/source_files_and_compilation.html#compiler-directives).

Not 100% sure if this is a bug or not - it's decided that the type of `b` is a C integer because there's only one assignment to it and it's clear that it can't overflow. However clearly in this case it does behave differently which the documentation for the default type inference suggests that it shouldn't do.

**Comment by scoder** [comment]
Hmm, yeah, difficult choice. If there is a chance that a variable is unassigned during type inference, it'd better not be inferred as a C type. OTOH, there will always be cases where we cannot determine safely whether a variable has been assigned in all valid(!) code paths, so this would probably reduce the usefulness of type inference.

That said, it's obviously a bug in the code as well to use a variable that has not been initialised. Cython warns about it at compile time, and users should better look at those warnings. So, I'm slightly leaning towards a "won't fix, at least for now", unless there is an actual realistic case in which this occurs.

@Zabolekar, could you provide some more details, or a longer snippet of your actual code in which this occurs? Is it a problem in valid code, or did you just stumble over it when fixing a bug in your code?

**Comment by Zabolekar** [comment]
@scoder You are right, I didn't actually want my code to raise `UnboundLocalError`, I've found it while fixing a bug. Cython compile-time warnings have been very helpful to me before, and it would be the best scenario here as well. However, this particular example produced no warning.

After some searching I've found the `warn.maybe_uninitialized` compiler directive. When it's set to True, I see the warning. Default is False. I'm slightly confused why the default is not True (if I understand you correctly, you would expect a warning as well), but, as far as I see, it's neither a bug nor undocumented.
